Deterministic flag in options wich initializes the random generator w…

a-mueller opened the pull request:
I added a deterministic flag to the options. If set the word cloud will always look the same given the same input.

Commit:
Deterministic flag in options wich initializes the random generator with a static seed.
This means the wordcloud will always look the same given the same input.

chrisrzhou commented on src/index.tsx:
```diff
@@ -101,6 +102,14 @@ function Wordcloud({
+      let randomGenerator = Math.random;
+
+      if (options.deterministic){
+        const seedrandom = require('seedrandom');
```
I'm okay with moving this to the top and incurring a slight hit.  You can use ES6 `import` syntax.

Once this is moved to the top, you can use `seedrandom` to initialize `randomGenerator`, so something like

```
const randomGenerator = options.deterministic ? seedrandom() : seedrandom('deterministic');
```

chrisrzhou commented on src/types.ts:
```diff
@@ -27,6 +27,7 @@ export interface Callbacks {
 export interface Options {
   colors: string[];
   enableTooltip: boolean;
+  deterministic: boolean;
```
small nit: put above row 29 (alphabetical)

chrisrzhou reviewed (commented):
This is a great suggestion and thank you for including it!

If you are adventurous, feel free to update the `docz` documentation for the `Options` MDX file.  Feel free to add a short section about using the `deterministic` option field. You can build the docs and test them locally with `yarn dev`.

I'll be making slight improvement to the typescript types this weekend, and reducing bundle size, and I can include your changes for a next version release soon!

Commit:
Added description to Options.mdx (couldn't test this however as yarn dev kept complaining about missing eliptic module)
Changed dependency for seeded random generator to pseudo-math-random instead of seedrandom
Passing the random generator to the rotate / color picking as well to keep them deterministic as well.

a-mueller commented:
Hey,
I made some of the changes you suggested. My typescript knowledge unfortunately was not enough to make the import work instead of using require. Kept having issues with default exports / namespace not being callable. I did extend the commit to include the rotate and colour choosing as well.

Commit:
Updated yarn.lock as react-rooter 4.0.0 caused a 404

chrisrzhou commented on src/index.tsx:
```diff
@@ -7,6 +7,7 @@ import render from './render';
 import { Callbacks, MinMaxPair, Options, Scale, Spiral, Word } from './types';
 import { getDefaultColors, getFontScale, getText, rotate } from './utils';
 
+
```
remove extra line

chrisrzhou commented on src/index.tsx:
```diff
@@ -101,6 +103,12 @@ function Wordcloud({
         .slice(0, maxWords);
 
+
+      const pmr = require('pseudo-math-random');
```
Move to top of file (maybe after line 8)

chrisrzhou commented on src/index.tsx:
```diff
@@ -101,6 +103,12 @@ function Wordcloud({
 
+
+      const pmr = require('pseudo-math-random');
+      const randomGenerator: ()=> number = options.deterministic ? pmr('deterministic') : pmr();
```
You can fix a few code formatting issues by running `yarn lint`.  It should fix up most of the formatting with the prettier rules in the codebase.

chrisrzhou replied:
You can probably call this `random`, since it effectively is a callback replacing `Math.random` in the code here.

chrisrzhou reviewed (approved):
Hey thanks a lot for updating the docs and generalizing `Math.random` calls with `pmr`.  This looks good to me.  Please address a few things before merging in

- rename `randomGenerator` with `random`
- Run `yarn lint` after code changes so `prettier` can auto format the codebase.
- With regards to the `yarn.lock` file and error you're seeing, I'm not sure if deleting the lock file and running `yarn` can help solve the problem.  I can help look into the issue in a future PR.

Thanks!

Commit:
rename randomGenerator -> random
code style changes

a-mueller commented:
Done.

Had to "eslint-disable-line" the pmr import after moving it to the suggested position. Didn't want to take the rule out the lint config alltogether but feel free to change that.

chrisrzhou commented:
Thanks so much for the changes here and enhancing the component!
